File: app/src/main/assets/linux/bin/atlas_cutter.py
#!/usr/bin/env python3
"""
UIPackedAtlas 图集切割工具
从 KTX 图集+Lua坐标元数据 → 切割出单个UI元素PNG，按前缀自动分类

用法:
  python3 atlas_cutter.py                           # 切割所有图集
  python3 atlas_cutter.py UIPackedAtlas1            # 只切割指定图集
  python3 atlas_cutter.py --dry-run                 # 预览，不实际切割
  python3 atlas_cutter.py --classify-only           # 只分类(不重新切割)
"""

import re
import os
import sys
import struct
import importlib.util
from collections import defaultdict
from pathlib import Path

# ── 导入转换模块 ──
spec = importlib.util.spec_from_file_location('converter', '/sdcard/MT2/mcp/转换_v2.py')
converter = importlib.util.module_from_spec(spec)
spec.loader.exec_module(converter)

from PIL import Image
import numpy as np

# ── 配置 ──
KTX_DIR = '/sdcard/MT2/apks/sky'
JSON_DIR = '/sdcard/MT2/mcp/json'
OUTPUT_BASE = '/sdcard/MT2/mcp/sliced_ui'

# Lua元数据文件
ATLAS_LUA_FILES = {
    'UIPackedAtlas.lua': 'UIPackedAtlas',
    'UIMapAtlas.lua': 'UiMapAtlas',
    'UiColorAtlas.lua': 'UiColorAtlas',
}

# ── 分类规则 ──
def classify(name):
    """根据名称前缀返回分类目录（光遇中文术语）"""
    rules = [
        ('UiEmote', '表情'),
        ('UiButton', '按键图标'),
        ('UiMisc', '杂项图标'),
        ('UiMap', '地图'),
        ('UiAnalog', '摇杆'),
        ('UiBorder', '边框'),
        ('UiBrand', '品牌'),
        ('UiCrab', '螃蟹'),
        ('UiDisplay', '显示'),
        ('UiMenu', '菜单'),
        ('UiHud', '界面'),
        ('UiOutfitBody', '装扮/裤子'),
        ('UiOutfitCape', '装扮/斗篷'),
        ('UiOutfitHair', '装扮/发型'),
        ('UiOutfitHorn', '装扮/头角'),
        ('UiOutfitMask', '装扮/面具'),
        ('UiOutfitProp', '装扮/道具'),
        ('UiOutfitFeet', '装扮/鞋子'),
        ('UiOutfitFace', '装扮/面部'),
        ('UiOutfitNeck', '装扮/颈饰'),
        ('UiOutfitHat', '装扮/头饰'),
        ('UiOutfit', '装扮/其他'),
        ('UiSocial', '社交'),
        ('UiSpirit', '先祖'),
        ('UiFriend', '好友'),
        ('UiGift', '礼物'),
        ('UiQuest', '任务'),
        ('UiSeason', '季节'),
        ('UiSettings', '设置'),
        ('UiSpell', '魔法'),
        ('UiWing', '光翼'),
        ('UiNPC', 'NPC'),
        ('UiMatch', '匹配'),
        ('UiToggle', '开关'),
        ('UiRadial', '径向渐变'),
        ('UiGradient', '渐变'),
        ('UiMusic', '音乐'),
        ('UiPlaceholder', '占位图'),
        ('UiWip', '开发中'),
    ]
    for prefix, folder in rules:
        if name.startswith(prefix):
            return folder
    return '其他'


# ── Lua解析 ──
def parse_atlas_lua(filepath):
    """解析 ImageRegion 定义，返回 [(name, atlas, u1, v1, u2, v2), ...]"""
    with open(filepath, 'r') as f:
        content = f.read()
    
    pattern = r'resource "ImageRegion" "([^"]+)" \{ image = "([^"]+)", uv = \{ ([^}]+) \} \}'
    matches = re.findall(pattern, content)
    
    regions = []
    for name, atlas, uv_str in matches:
        parts = [p.strip() for p in uv_str.split(',')]
        try:
            u1 = eval(parts[0])
            v1 = eval(parts[1])
            u2 = eval(parts[2])
            v2 = eval(parts[3])
            regions.append((name, atlas, u1, v1, u2, v2))
        except Exception as e:
            print(f"  [警告] 无法解析UV: {name}: {uv_str} - {e}")
    
    return regions


# ── 图集加载与解码 ──
def load_atlas(atlas_name):
    """加载KTX图集，返回 (width, height, rgba_pixels)"""
    ktx_path = os.path.join(KTX_DIR, f'{atlas_name}.ktx')
    if not os.path.exists(ktx_path):
        print(f"  [错误] KTX文件不存在: {ktx_path}")
        return None, None, None
    
    with open(ktx_path, 'rb') as f:
        ktx_data = f.read()
    
    try:
        w, h, pixels, fmt_name = converter.ktx_to_png(ktx_data)
        print(f"  加载: {atlas_name}.ktx → {fmt_name}, {w}×{h}")
        
        # 如果是R11_EAC灰度图集，应用模式2视觉增强
        if 'R11_EAC' in fmt_name:
            print(f"    应用视觉增强 (模式2)...")
            pixels = converter.enhance_grayscale(pixels, w, h)
        
        return w, h, pixels
    except Exception as e:
        print(f"  [错误] 解码失败: {e}")
        return None, None, None


# ── 切割核心 ──
def crop_region(pixels, atlas_w, atlas_h, u1, v1, u2, v2):
    """从图集像素中裁剪指定UV区域，返回 (cropped_pixels, w, h)"""
    x1 = int(round(u1 * atlas_w))
    y1 = int(round(v1 * atlas_h))
    x2 = int(round(u2 * atlas_w))
    y2 = int(round(v2 * atlas_h))
    
    # 边界保护
    x1 = max(0, min(x1, atlas_w))
    y1 = max(0, min(y1, atlas_h))
    x2 = max(0, min(x2, atlas_w))
    y2 = max(0, min(y2, atlas_h))
    
    w = x2 - x1
    h = y2 - y1
    
    if w <= 0 or h <= 0:
        return None, 0, 0
    
    arr = np.frombuffer(bytes(pixels), dtype=np.uint8).reshape(atlas_h, atlas_w, 4)
    crop = arr[y1:y2, x1:x2].copy()
    
    return bytearray(crop.tobytes()), w, h


# ── 主流程 ──
def main():
    dry_run = '--dry-run' in sys.argv
    classify_only = '--classify-only' in sys.argv
    target_atlas = None
    
    for arg in sys.argv[1:]:
        if not arg.startswith('--') and arg not in ('1', '2'):
            target_atlas = arg
    
    print("=" * 60)
    print("UIPackedAtlas 图集切割工具")
    print("=" * 60)
    
    # 1. 解析所有Lua元数据
    all_regions = []
    for lua_file, atlas_prefix in ATLAS_LUA_FILES.items():
        lua_path = os.path.join(JSON_DIR, lua_file)
        if not os.path.exists(lua_path):
            continue
        regions = parse_atlas_lua(lua_path)
        print(f"\n解析 {lua_file}: {len(regions)} 个区域")
        all_regions.extend(regions)
    
    print(f"\n总计: {len(all_regions)} 个UI元素")
    
    # 2. 按图集分组
    atlas_groups = defaultdict(list)
    for name, atlas, u1, v1, u2, v2 in all_regions:
        if target_atlas and atlas != target_atlas:
            continue
        atlas_groups[atlas].append((name, u1, v1, u2, v2))
    
    if target_atlas:
        print(f"筛选目标图集: {target_atlas}")
    print(f"待处理图集: {len(atlas_groups)} 个")
    
    # 3. 逐图集处理
    total_cropped = 0
    total_skipped = 0
    stats = defaultdict(lambda: {'count': 0, 'total_px': 0})
    
    for atlas_name in sorted(atlas_groups.keys()):
        regions = atlas_groups[atlas_name]
        print(f"\n{'─' * 50}")
        print(f"处理: {atlas_name} ({len(regions)} 个区域)")
        
        if dry_run:
            for name, u1, v1, u2, v2 in regions[:5]:
                print(f"  [预览] {name}: UV=({u1:.4f},{v1:.4f})-({u2:.4f},{v2:.4f})")
            if len(regions) > 5:
                print(f"  ... 还有 {len(regions)-5} 个")
            total_cropped += len(regions)
            continue
        
        if classify_only:
            # 只做分类统计
            for name, u1, v1, u2, v2 in regions:
                cat = classify(name)
                stats[cat]['count'] += 1
            continue
        
        # 加载图集
        atlas_w, atlas_h, pixels = load_atlas(atlas_name)
        if pixels is None:
            total_skipped += len(regions)
            continue
        
        # 切割每个区域
        for name, u1, v1, u2, v2 in regions:
            crop_data, cw, ch = crop_region(pixels, atlas_w, atlas_h, u1, v1, u2, v2)
            
            if crop_data is None:
                total_skipped += 1
                continue
            
            # 分类目录
            cat = classify(name)
            out_dir = os.path.join(OUTPUT_BASE, cat)
            os.makedirs(out_dir, exist_ok=True)
            
            # 保存PNG
            out_path = os.path.join(out_dir, f'{name}.png')
            img = Image.frombytes('RGBA', (cw, ch), bytes(crop_data))
            img.save(out_path, 'PNG')
            
            total_cropped += 1
            stats[cat]['count'] += 1
            stats[cat]['total_px'] += cw * ch
    
    # 4. 输出统计
    print(f"\n{'=' * 60}")
    print(f"完成!")
    print(f"  切割成功: {total_cropped}")
    print(f"  跳过: {total_skipped}")
    print(f"  输出目录: {OUTPUT_BASE}")
    
    if stats:
        print(f"\n分类统计:")
        for cat in sorted(stats.keys()):
            s = stats[cat]
            print(f"  {cat:<25} {s['count']:>5} 个  ({s['total_px']:>12,} px²)")
        print(f"  {'总计':<25} {total_cropped:>5} 个")


if __name__ == '__main__':
    main()

File: app/src/main/java/com/ktxconverter/app/AtlasCutter.kt
package com.ktxconverter.app

/**
 * UIPackedAtlas 图集切割工具
 *
 * 从 KTX 图集 + Lua坐标元数据 → 切割出单个UI元素PNG，按前缀自动分类
 * 移植自 atlas_cutter.py
 *
 * 分类规则基于光遇中文UI术语
 */
object AtlasCutter {

    data class ImageRegion(
        val name: String,
        val atlas: String,
        val u1: Float,
        val v1: Float,
        val u2: Float,
        val v2: Float
    )

    data class CropResult(
        val name: String,
        val category: String,
        val pixels: ByteArray,
        val width: Int,
        val height: Int
    )

    /**
     * 解析 Lua 中的 ImageRegion 定义
     * 格式: resource "ImageRegion" "Name" { image = "AtlasName", uv = { u1, v1, u2, v2 } }
     */
    private val regionPattern = Regex(
        """resource\s+"ImageRegion"\s+"([^"]+)"\s*\{\s*image\s*=\s*"([^"]+)",\s*uv\s*=\s*\{\s*([^}]+)\s*\}\s*\}"""
    )

    fun parseAtlasLua(content: String): List<ImageRegion> {
        return regionPattern.findAll(content).mapNotNull { match ->
            try {
                val name = match.groupValues[1]
                val atlas = match.groupValues[2]
                val uvStr = match.groupValues[3]
                val parts = uvStr.split(",").map { it.trim() }
                if (parts.size >= 4) {
                    val u1 = evalFloat(parts[0])
                    val v1 = evalFloat(parts[1])
                    val u2 = evalFloat(parts[2])
                    val v2 = evalFloat(parts[3])
                    ImageRegion(name, atlas, u1, v1, u2, v2)
                } else null
            } catch (e: Exception) {
                null
            }
        }.toList()
    }

    /**
     * 从图集像素中裁剪指定 UV 区域（带边缘精修）
     */
    fun cropRegion(
        pixels: ByteArray, atlasW: Int, atlasH: Int,
        u1: Float, v1: Float, u2: Float, v2: Float,
        skipSharpen: Boolean = false
    ): CropData? {
        // 用 floor/ceil 向外扩展2px 捕获抗锯齿边缘（ETC2块压缩需要更大边距）
        val pad = 2
        val x1 = kotlin.math.floor(u1 * atlasW).toInt().coerceIn(0, atlasW - 1) - pad
        val y1 = kotlin.math.floor(v1 * atlasH).toInt().coerceIn(0, atlasH - 1) - pad
        val x2 = kotlin.math.ceil(u2 * atlasW).toInt().coerceIn(1, atlasW) + pad
        val y2 = kotlin.math.ceil(v2 * atlasH).toInt().coerceIn(1, atlasH) + pad

        val cx1 = x1.coerceAtLeast(0)
        val cy1 = y1.coerceAtLeast(0)
        val cx2 = x2.coerceAtMost(atlasW)
        val cy2 = y2.coerceAtMost(atlasH)

        val w = cx2 - cx1
        val h = cy2 - cy1
        if (w <= 0 || h <= 0) return null

        val cropped = ByteArray(w * h * 4)
        for (y in 0 until h) {
            val srcRow = (cy1 + y) * atlasW * 4 + cx1 * 4
            val dstRow = y * w * 4
            System.arraycopy(pixels, srcRow, cropped, dstRow, w * 4)
        }

        // 去块 + 锐化（保持边缘清晰）；ENHANCE 模式下跳过（已有 blur+二值化）
        val processed = if (skipSharpen) cropped else sharpenEdges(cropped, w, h)
        // 裁剪透明边距
        return trimTransparent(processed, w, h)
    }

    /**
     * 边缘锐化：反锐化掩模，保持边缘清晰
     * 对 ETC2 块压缩导致的模糊做小幅锐化
     */
    private fun sharpenEdges(pixels: ByteArray, w: Int, h: Int): ByteArray {
        val result = pixels.copyOf()
        val amount = 0.25f  // 轻量锐化，避免边缘光晕
        for (y in 1 until h - 1) {
            for (x in 1 until w - 1) {
                val idx = (y * w + x) * 4
                val a = result[idx + 3].toInt() and 0xFF
                // 只处理完全不透明的内部像素，保护边缘渐变
                if (a <= 200) continue
                // 对 RGB 三通道做反锐化掩模
                for (c in 0..2) {
                    val center = result[idx + c].toInt() and 0xFF
                    // 4邻域均值
                    val avg = (
                        (result[((y-1)*w + x) * 4 + c].toInt() and 0xFF) +
                        (result[((y+1)*w + x) * 4 + c].toInt() and 0xFF) +
                        (result[(y*w + x-1) * 4 + c].toInt() and 0xFF) +
                        (result[(y*w + x+1) * 4 + c].toInt() and 0xFF)
                    ) / 4
                    // unsharp: result = center + (center - avg) * amount
                    val sharp = center + ((center - avg) * amount).toInt()
                    result[idx + c] = sharp.coerceIn(0, 255).toByte()
                }
            }
        }
        return result
    }

    /**
     * 裁剪四周全透明像素（保留1px边距防止再次出现硬边）
     */
    private fun trimTransparent(pixels: ByteArray, w: Int, h: Int): CropData {
        var left = 0
        var top = 0
        var right = w
        var bottom = h

        // 从上往下找第一个不透明行
        for (y in 0 until h) {
            if (!rowAllTransparent(pixels, w, y)) { top = y; break }
        }
        // 从下往上
        for (y in h - 1 downTo 0) {
            if (!rowAllTransparent(pixels, w, y)) { bottom = y + 1; break }
        }
        // 从左往右
        for (x in 0 until w) {
            if (!colAllTransparent(pixels, w, h, x)) { left = x; break }
        }
        // 从右往左
        for (x in w - 1 downTo 0) {
            if (!colAllTransparent(pixels, w, h, x)) { right = x + 1; break }
        }

        // 保留2px边距（与裁切边距一致）
        left = (left - 2).coerceAtLeast(0)
        top = (top - 2).coerceAtLeast(0)
        right = (right + 2).coerceAtMost(w)
        bottom = (bottom + 2).coerceAtMost(h)

        val nw = right - left
        val nh = bottom - top
        if (nw <= 0 || nh <= 0) return CropData(pixels, w, h)

        val trimmed = ByteArray(nw * nh * 4)
        for (y in 0 until nh) {
            val srcRow = (top + y) * w * 4 + left * 4
            System.arraycopy(pixels, srcRow, trimmed, y * nw * 4, nw * 4)
        }
        return CropData(trimmed, nw, nh)
    }

    private fun rowAllTransparent(pixels: ByteArray, w: Int, y: Int): Boolean {
        val rowStart = y * w * 4
        for (x in 0 until w) {
            if ((pixels[rowStart + x * 4 + 3].toInt() and 0xFF) > 0) return false
        }
        return true
    }

    private fun colAllTransparent(pixels: ByteArray, w: Int, h: Int, x: Int): Boolean {
        for (y in 0 until h) {
            if ((pixels[(y * w + x) * 4 + 3].toInt() and 0xFF) > 0) return false
        }
        return true
    }

    /**
     * 按名称前缀分类（光遇中文UI术语）
     * 分类名中的 / 替换为 _ 避免产生嵌套子目录
     */
    fun classify(name: String): String {
        val rules = listOf(
            "UiEmote" to "表情",
            "UiButton" to "按键图标",
            "UiMisc" to "杂项图标",
            "UiLightBulb" to "杂项图标",
            "UiMap" to "地图",
            "UiAnalog" to "摇杆",
            "UiBorder" to "边框",
            "UiBrand" to "品牌",
            "UiCrab" to "螃蟹",
            "UiDisplay" to "显示",
            "UiFreshman" to "新手引导",
            "UiMenu" to "菜单",
            "UiHud" to "界面",
            "UiOutfitBody" to "装扮_裤子",
            "UiOutfitCape" to "装扮_斗篷",
            "UiOutfitHair" to "装扮_发型",
            "UiOutfitHorn" to "装扮_头角",
            "UiOutfitMask" to "装扮_面具",
            "UiOutfitProp" to "装扮_道具",
            "UiOutfitFeet" to "装扮_鞋子",
            "UiOutfitFace" to "装扮_面部",
            "UiOutfitNeck" to "装扮_颈饰",
            "UiOutfitHat" to "装扮_头饰",
            "UiOutfitTail" to "装扮_尾巴",
            "UiOutfitWing" to "装扮_翅膀",
            "UiOutfitNone" to "装扮_占位",
            "UiOutfit" to "装扮_其他",
            "UiPersonality" to "性格图标",
            "UiSharedSpace" to "共享空间",
            "UiSocial" to "社交",
            "UiSeason" to "季节",
            "UiMatch" to "匹配",
            "UiToggle" to "开关",
            "UiRadial" to "径向渐变",
            "UiGradient" to "渐变",
            "UiMusic" to "音乐",
            "UiPlaceholder" to "占位图",
            "UiWip" to "开发中",
        )
        for ((prefix, folder) in rules) {
            if (name.startsWith(prefix)) return folder
        }
        return "其他"
    }

    data class CropData(
        val pixels: ByteArray,
        val width: Int,
        val height: Int
    )

    /**
     * 解析浮点数（支持分数形式 1/2048 和普通小数 0.0625）
     */
    private fun evalFloat(expr: String): Float {
        return try {
            if (expr.contains("/")) {
                val parts = expr.split("/")
                parts[0].trim().toFloat() / parts[1].trim().toFloat()
            } else {
                expr.toFloat()
            }
        } catch (e: Exception) {
            0f
        }
    }
}